flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An Estimator which implements the KNN algorithm.
+ *
+ * <p>See: https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm.
+ */
+public class Knn implements Estimator<Knn, KnnModel>, KnnParams<Knn> {
+
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public Knn() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public KnnModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        /* Tuple3 : <feature, label, norm> */
+        DataStream<Tuple3<DenseVector, Double, Double>> inputDataWithNorm =
+                computeNorm(tEnv.toDataStream(inputs[0]));
+        DataStream<KnnModelData> modelData = genModelData(inputDataWithNorm);
+        KnnModel model = new KnnModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static Knn load(StreamExecutionEnvironment env, String path) throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /**
+     * Generates knn model data.
+     *
+     * @param inputDataWithNorm Input data with feature norm.
+     * @return Knn model.
+     */
+    private static DataStream<KnnModelData> genModelData(
+            DataStream<Tuple3<DenseVector, Double, Double>> inputDataWithNorm) {
+        DataStream<KnnModelData> modelData =
+                DataStreamUtils.mapPartition(
+                        inputDataWithNorm,
+                        new RichMapPartitionFunction<
+                                Tuple3<DenseVector, Double, Double>, KnnModelData>() {
+                            @Override
+                            public void mapPartition(
+                                    Iterable<Tuple3<DenseVector, Double, Double>> values,
+                                    Collector<KnnModelData> out) {
+                                List<Tuple3<DenseVector, Double, Double>> buffer =
+                                        new ArrayList<>(1);
+                                for (Tuple3<DenseVector, Double, Double> value : values) {
```

Comment:
nits: Can we do the following renaming?

`value` -> `dataPoint`
`buffer` -> `bufferedDataPoints`
`values` -> `dataPoints`

Reply:
OK